z3c/dependencychecker/db.py

```diff
@@ -43,3 +48,186 @@ def _filter_duplicates(self, imports):
+    def get_unneeded_requirements(self):
+        all_but_test_requirements = self._requirements.copy()
+        for extra in self._extras_requirements:
+            all_but_test_requirements.update(self._extras_requirements[extra])
+        test_requirements = self._get_test_extra()
+        if test_requirements:
+            for dotted_name in test_requirements:
+                all_but_test_requirements.remove(dotted_name)
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(all_but_test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def requirements_that_should_be_test_requirements(self):
+        filters = (
+            self._filter_testing_imports,
+            self._filter_python_standard_library,
+            self._filter_own_package,
+        )
+        non_testing_imports = self._process_pipeline(
+            self.imports_used,
+            filters,
+        )
+        requirements_not_used = [
+            x
+            for x in self._requirements
+            if self._filter_requirements_lists(non_testing_imports, x)
+        ]
+        filters = (
+            self._filter_only_testing_imports,
+            self._filter_python_standard_library,
+            self._filter_own_package,
+        )
+        testing_imports = self._process_pipeline(
+            self.imports_used,
+            filters,
+        )
+        should_be_test_requirements = [
+            x
+            for x in requirements_not_used
+            if not self._filter_requirements_lists(testing_imports, x)
+        ]
+        unique = self._get_unique_imports(
+            imports_list=should_be_test_requirements,
+        )
+        return unique
+
+    def get_unneeded_test_requirements(self):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return []
+
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def _process_pipeline(self, objects, filters):
+        final_list = []
+        for single_object in objects:
+            for filter_func in filters:
+                if not filter_func(single_object):
+                    break
+            else:
+                final_list.append(single_object)
+
+        return final_list
+
+    def _get_unique_imports(self, imports_list=None):
+        if imports_list is None:
+            imports_list = self.imports_used
+        unique_dotted_names = list(set(imports_list))
+        sorted_unique_dotted_names = sorted(unique_dotted_names)
```

Comment:
Sorting them is neat and tidy!

Reply:
And makes tests easier to reproduce and assert 🙂